Content.Shared/Containers/ItemSlot/ItemSlotsSystem.cs

```diff
@@ -200,19 +208,41 @@ private void OnInteractUsing(EntityUid uid, ItemSlotsComponent itemSlots, Intera
                 return;
 
             var slots = new List<ItemSlot>();
+            // if itemSlots.Slots.Count == 0, we wouldn't want to show the locked popup that would be weird
+            var allLocked = itemSlots.Slots.Count != 0;
+            ItemSlot? lastFailedSlot = null;
             foreach (var slot in itemSlots.Slots.Values)
             {
                 if (!slot.InsertOnInteract)
                     continue;
 
-                if (!CanInsert(uid, args.Used, args.User, slot, swap: slot.Swap, popup: args.User))
+                if (!CanInsert(uid, args.Used, args.User, slot, slot.Swap))
+                {
+                    if (!slot.Locked)
+                        allLocked = false;
+                    lastFailedSlot = slot;
                     continue;
+                }
 
                 slots.Add(slot);
             }
 
             if (slots.Count == 0)
+            {
+                // no slots failed, so uh,
+                if (lastFailedSlot == null)
+                    return;
+
+                // it's a bit weird that the popupMessage is stored with the item slots themselves, but in practice
+                // the popup messages will just all be the same, so it's probably fine.
+                //
+                // doing a check to make sure that they're all the same or something is probably frivolous
+                if (allLocked && lastFailedSlot.LockedFailPopup != null)
+                    _popupSystem.PopupClient(Loc.GetString(lastFailedSlot.LockedFailPopup), uid, args.User);
+                if (!allLocked && lastFailedSlot.WhitelistFailPopup != null)
+                    _popupSystem.PopupClient(Loc.GetString(lastFailedSlot.WhitelistFailPopup), uid, args.User);
```

Comment:
Alternative solution: if the item is not whitelisted, just show the popup for that. You don't need to worry about whether the slots are locked or not if you would never be able to insert the item in the first place. Otherwise, maybe show the locked popup.

Reply:
Seems like that'd be fine.